umbral/keys.py

```diff
@@ -38,6 +42,26 @@ def gen_key(cls, params: UmbralParameters=None):
         bn_key = BigNum.gen_rand(params.curve)
         return cls(bn_key, params)
 
+    @classmethod
+    def derive_key_from_label(cls, master_secret: bytes, label: bytes, 
```

Comment:
Another option is to move this `derive_key_from_label` method onto `UmbralPrivateKey`.

This way, a user can load up a "master key" they have stored and derive keys from labels in a more simplified API on the key.

Reply:
While I completely agree that we should make this as misuse resistance as possible and the problems associated to an arbitrary bytes input, I think that allowing UmbralPrivateKeys to be used as master secret to other derived keys is relatively dangerous. The problem I see is that if we allow to use the "master" UmbralPrivateKey to be used somewhere (e.g., to create KFrags), then in the event of a leakage or a collusion attack, all the derived keys are compromised too. 

So, some alternatives:
1) Creating a MasterSecretKey class to represent a high-entropy random sequence of bytes, and use this to derive UmbralPrivateKeys

2) Creating a special subclass of UmbralPrivateKey that somehow doesn't allow to be used by Umbral directly

Thoughts?

Reply:
@cygnusv: This pattern sounds good to me.

However, I have never liked the name "MasterKey" (or in this case, "MasterSecretKey") because, even in a world where so many things are called "keys" even though they haven't even a vaguely coherent metaphor with an unlocking device, this one is particularly stretched.  If it will never be used as a key, how about:

```
class MasterSecret(UmbralPrivateKey):
    ...
```

Reply:
Oh yes, you are right about the name. I'm probably too promiscuous with the work "key".
Now that I think of, it probably has more sense that MasterSecret (or Secret or whatever) is a superclass...

Reply:
Yeah?  But isn't this what we want the implementer not to do?  Lest they re-use a key per @cygnusv's concern above?

Reply:
OK, I see there are two approaches here:

1) There is a conversion path between `MasterSecret` and `UmbralPrivateKey` (in one direction, the other, or both). This not only includes explicit methods (e.g., `UmbralPrivateKey.to_master_secret()`) but also inheritance relations (e.g., if `MasterSecret` is subclass of `UmbralPrivateKey`, then trivially a `MasterSecret` can already be used as `UmbralPrivateKey`). The problem with this approach is that if there's a way to use the same secret material as an `UmbralPrivateKey`, then we have the problems I mentioned before in case of key leakage. I don't see a clear advantage of this approach, apart from the fact that the users only need to care about one of the classes. 

2) There is no such conversion path. If we go in this direction, then there is no way for the implementer to reuse secret material between `MasterSecrets` and `UmbralPrivateKeys`. This is probably the safest approach.

Reply:
I don't see any advantages of the former.  If there aren't any, let's go with the latter.